libyul/optimiser/MemoryStoreRemover.h

```diff
@@ -0,0 +1,88 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Optimizer step that removes `mstore(x, y)` if the memory location `[x, x + 32)` is never read.
+ */
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Solver.h>
+
+#include <libyul/AST.h>
+#include <libyul/Dialect.h>
+
+namespace solidity::yul
+{
+
+/**
+ * Optimization step that removes `mstore(x, y)` if the memory location `[x, x + 32)` is never used.
```

Comment:
I actually did not realize this until now: For memory, you can erase all stores that are never read from, for storage you can only erase those that are overwritten!